src/cargo/ops/cargo_generate_lockfile.rs

```diff
@@ -108,4 +134,47 @@ pub fn update_lockfile(manifest_path: &Path,
             None => {}
         }
     }
+
+    fn compare_dependency_graphs<'a>(previous_resolve: &'a Resolve,
+                                     resolve: &'a Resolve) ->
+                                     Vec<(Vec<&'a PackageId>, Vec<&'a PackageId>)> {
+        // Map (package name, package source) to (removed versions, added versions).
+        fn changes_key<'a>(dep: &'a PackageId) -> (&'a str, &'a SourceId) {
+            (dep.name(), dep.source_id())
+        }
+
+        fn vec_subtract<'a, T>(a: &'a Vec<T>, b: &'a Vec<T>) -> Vec<T>
+            where T: Ord + Clone {
+            let mut result: Vec<T> = a.clone();
+            let mut b: Vec<T> = b.clone();
+            b.sort();
+            result.retain(|x| b.binary_search(x).is_err());
+            result
+        }
+
+        let mut changes = HashMap::new();
+
+        for dep in previous_resolve.iter() {
+            changes.insert(changes_key(dep), (vec![dep], vec![]));
+        }
+        for dep in resolve.iter() {
+            let (_, ref mut added) = *changes.entry(changes_key(dep))
+                                             .or_insert_with(|| (vec![], vec![]));
+            added.push(dep);
+        }
+
+        for (_, v) in changes.iter_mut() {
+            let (ref mut old, ref mut new) = *v;
+            let removed = vec_subtract(old, new);
+            let added = vec_subtract(new, old);
+            *old = removed;
+            *new = added;
+        }
+
+        // Sort the packages by their names.
+        let mut packages: Vec<(&'a str, &'a SourceId)> =
+            changes.keys().map(|x| *x).collect();
+        packages.sort_by(|a, b| a.0.cmp(b.0));
```

Comment:
Could this just call `sort`? That should naturally handle the orderings of names while also sorting by the sources next.